src/watcher/watcher.ts

```diff
@@ -40,28 +41,62 @@ export function createWatcher(opts: Opts) {
 
   const pluginWatchContributions = opts.plugins.reduce(
     (patterns, p) =>
-      patterns.concat(p.dev.addToSettings.watchFilePatterns || []),
+      patterns.concat(p.dev.addToWatcherSettings.watchFilePatterns ?? []),
     [] as string[]
   )
 
   const pluginIgnoreContributions = opts.plugins.reduce(
     (patterns, p) =>
-      patterns.concat(p.dev.addToSettings.ignoreFilePatterns || []),
+      patterns.concat(
+        p.dev.addToWatcherSettings.core?.ignoreFilePatterns ?? []
+      ),
     [] as string[]
   )
+  const isIgnoredByCoreListener = createPathMatcher({
+    toMatch: pluginIgnoreContributions,
+  })
 
   const watcher = watch([opts.layout.sourceRoot, ...pluginWatchContributions], {
-    ignored: ['./node_modules', './.*', ...pluginIgnoreContributions],
+    ignored: ['./node_modules', './.*'],
     ignoreInitial: true,
-    onAll(event, file) {
-      for (const p of opts.plugins) {
-        p.dev.onFileWatcherEvent?.(event, file)
-      }
-      restartRunner(file)
-    },
     cwd: process.cwd(), // prevent globbed files and required files from being watched twice
   })
 
+  /**
+   * Core watcher listener
+   */
+  watcher.on('all', (_event, file) => {
+    if (isIgnoredByCoreListener(file)) {
+      return log('global watcher - DID NOT match file: %s', file)
+    }
+
+    log('global watcher - matched file: %s', file)
+    restartRunner(file)
+  })
+
+  /**
+   * Plugins watcher listeners
+   */
+  for (const p of opts.plugins) {
+    if (p.dev.onFileWatcherEvent) {
+      const isMatchedByPluginListener = createPathMatcher({
+        toMatch: p.dev.addToWatcherSettings.plugin?.allowFilePatterns,
+        toIgnore: p.dev.addToWatcherSettings.plugin?.ignoreFilePatterns,
+      })
+
+      watcher.on('all', (event, file, stats) => {
```

Comment:
I think we can get by with a single `.on` registration with our present system.

IIUC by sharing `watcher` instance it means that the ignored files in core watcher affect what plugins will be able to see. Certainly something to think more about in the future as it means plugins can break each other in subtle ways. But then, we could consider core ignore additions as `WARNING ADVANCED FEATURE MAKE SURE YOU KNOW WHAT YOU'RE DOING` kind of thing. We'll see :)

Reply:
> IIUC by sharing watcher instance it means that the ignored files in core watcher affect what plugins will be able to see

Related to https://github.com/prisma-labs/graphql-santa/pull/279#discussion_r366958202. No ignored files in core watcher **will not** after what plugins can see for the reasons mentioned in the comment (and the comment linked in that comment 🙈)

Reply:
> I think we can get by with a single .on registration with our present system.

Agreed 👍. Will add a TODO for now if you don't mind so that I can move on